lib/collection/src/shards/local_shard/clock_map.rs

```diff
@@ -362,4 +374,217 @@ mod tests {
 
         assert_eq!(input, output);
     }
+
+    #[rstest]
+    #[case::with_empty_clock_map(Helper::empty())]
+    #[case::with_clock_map_at_tick_0(Helper::at_tick_0())]
+    fn clock_map_accept_tick_0(#[case] mut helper: Helper) {
+        // Accept tick `0`
+        helper.advance(tag(0)).assert(true, 0);
+    }
+
+    #[rstest]
+    #[case::with_empty_clock_map(Helper::empty())]
+    #[case::with_clock_map_at_tick_0(Helper::at_tick_0())]
+    fn clock_map_advance_to_next_tick(#[case] mut helper: Helper) {
+        // Advance to the next tick
+        for tick in 1..10 {
+            helper.advance(tag(tick)).assert(true, tick);
+        }
+    }
+
+    #[rstest]
+    #[case::with_empty_clock_map(Helper::empty())]
+    #[case::with_clock_map_at_tick_0(Helper::at_tick_0())]
+    fn clock_map_advance_to_newer_tick(#[case] mut helper: Helper) {
+        // Advance to a newer tick
+        for tick in [10, 20, 30, 40, 50] {
+            helper.advance(tag(tick)).assert(true, tick);
+        }
+    }
+
+    #[test]
+    fn clock_map_accept_tick_0_with_non_empty_clock_map() {
+        let mut helper = Helper::default();
+
+        for tick in [10, 20, 30, 40, 50] {
+            // Advance to a non-zero tick (already tested in `clock_map_advance_to_newer_tick`)
+            helper.advance(tag(tick));
+
+            // Accept tick `0`
+            helper.advance(tag(0)).assert(true, tick);
+        }
+    }
+
+    #[test]
+    fn clock_map_reject_older_or_current_tick() {
+        let mut helper = Helper::default();
+
+        // Advance to a newer tick (already tested in `clock_map_advance_to_newer_tick`)
+        helper.advance(tag(10));
+
+        // Reject older tick
+        //
+        // Start from tick `1`, cause tick `0` is a special case that is always accepted
+        for older_tick in 1..10 {
+            helper.advance(tag(older_tick)).assert(false, 10);
+        }
+
+        // Reject current tick
+        for current_tick in [10, 10, 10, 10, 10] {
+            helper.advance(tag(current_tick)).assert(false, 10);
+        }
```

Comment:
If rejection means we won't write the operation to WAL, I'm not sure if we should do this.

What if we do want to rewrite an operation, since they're idempotent, it shouldn't matter.

Or are all cases where we might do this covered with force=true?

Reply:
> If rejection means we won't write the operation to WAL, I'm not sure if we should do this.

Do you mean older ticks or current ticks?

I think we strictly **cannot** write *older* ticks, and we *maybe* can write *current* ticks, but I'm not sure if we should.

If you just allow writing arbitrary older ticks into WAL, you fuck up the diff resolution.

> What if we do want to rewrite an operation, since they're idempotent, it shouldn't matter.

I don't think that's something we ever do? (Or maybe I don't understand what you mean here.)

> Or are all cases where we might do this covered with force=true?

Yeah, I think they are.

We generally want to keep WAL strictly sequential (with a small exception for `0` ticks), so we explicitly only accept *newer* ticks.

Slightly artificial example: what if ops 4 and 5 arrive from forward proxy, and then we (somehow) receive very late op 4 from original shard (and op 5 is lost forever). And now we have 4, 5, 4 in WAL and diff resolution is broken.

And we have `force = true` for rare cases when we want to "force" some operations into WAL, but in such cases we *have to* make sure these "forced" operations won't break WAL properties that we rely on.

Reply:
> We generally want to keep WAL strictly sequential (with a small exception for 0 ticks)...

BTW, I think our diff resolution can't handle this case yet, cause when shard is restarted it will send op(s) with tick `0`, and these ops will be written into WALs with *current* tick, and so we might have multiple different ops saved in WAL with the same tick, so we'll have to pick all of them.

Reply:
Alright. Let's merge and see what happens! :+1: